### PR TITLE
opdi/egeria#5407 - Restart KafaOpenMetadataEventConsumer thread if it does unexpectedly

### PR DESCRIPTION
**Pull request by jnhagelberg**
Signed-off-by: Jeff Hagelberg <jnhagelberg@us.ibm.com>

See odpi/egeria#5407

### CODE REVIEW

**Comment by planetf1** [comment]
@wbittles you may want to review?

File: open-metadata-implementation/adapters/open-connectors/event-bus-connectors/open-metadata-topic-connectors/kafka-open-metadata-topic-connector/src/main/java/org/odpi/openmetadata/adapters/eventbus/topic/kafka/KafkaOpenMetadataTopicConnector.java [hunk 1]
```diff
@@ -58,15 +64,32 @@
...
+    
+    private class KafkaConsumerExecutor extends ThreadPoolExecutor {
+        KafkaConsumerExecutor() {
+            super(1, 1, Long.MAX_VALUE, TimeUnit.MICROSECONDS, new LinkedBlockingQueue<>(1));
+        }
+
+
         @Override
         public void afterExecute(Runnable r, Throwable t) {
             super.afterExecute(r, t);
```
Comment (mandy-chessell) [inline open-metadata-implementation/adapters/open-connectors/event-bus-connectors/open-metadata-topic-connectors/kafka-open-metadata-topic-connector/src/main/java/org/odpi/openmetadata/adapters/eventbus/topic/kafka/KafkaOpenMetadataTopicConnector.java:87]: What happens if the server is shutting down - should it check that the connector is still active before restarting the threads?

**Review by mandy-chessell — CHANGES_REQUESTED** [review]
This may be ok but I can not see any logic that checks that the connector has is still active before restarting the threads.  Please confirm that this case is handled.

**Comment by jnhagelberg** [comment]
@mandy-chessell This uses the exact same logic that is used to restart the producer thread.  If the server is shutting down, the thread will be killed by the JVM, so that should not be an issue.  I will add logic to check whether or not the connector is active before restarting the thread.

**Comment by jnhagelberg** [comment]
@mandy-chessell I have pushed changes to address your review comments.

File: open-metadata-implementation/adapters/open-connectors/event-bus-connectors/open-metadata-topic-connectors/kafka-open-metadata-topic-connector/src/main/java/org/odpi/openmetadata/adapters/eventbus/topic/kafka/KafkaOpenMetadataTopicConnector.java (hunk 1 above)
Comment (jnhagelberg) [reply open-metadata-implementation/adapters/open-connectors/event-bus-connectors/open-metadata-topic-connectors/kafka-open-metadata-topic-connector/src/main/java/org/odpi/openmetadata/adapters/eventbus/topic/kafka/KafkaOpenMetadataTopicConnector.java:89]: Fixed

**Comment by mandy-chessell** [comment]
> @mandy-chessell This uses the exact same logic that is used to restart the producer thread. If the server is shutting down, the thread will be killed by the JVM, so that should not be an issue. I will add logic to check whether or not the connector is active before restarting the thread.

The JVM does not exit when the server shuts down.  The JVM only exits when the platform shuts down.  If the producer thread is restarted without checking whether the connector is active or not then it is wrong too.

File: open-metadata-implementation/adapters/open-connectors/event-bus-connectors/open-metadata-topic-connectors/kafka-open-metadata-topic-connector/src/main/java/org/odpi/openmetadata/adapters/eventbus/topic/kafka/KafkaOpenMetadataTopicConnector.java (hunk 1 above)
Comment (mandy-chessell) [reply open-metadata-implementation/adapters/open-connectors/event-bus-connectors/open-metadata-topic-connectors/kafka-open-metadata-topic-connector/src/main/java/org/odpi/openmetadata/adapters/eventbus/topic/kafka/KafkaOpenMetadataTopicConnector.java:89]: Just need to update the comment so it matches the code and then I think it is ok

Comment (jnhagelberg) [reply open-metadata-implementation/adapters/open-connectors/event-bus-connectors/open-metadata-topic-connectors/kafka-open-metadata-topic-connector/src/main/java/org/odpi/openmetadata/adapters/eventbus/topic/kafka/KafkaOpenMetadataTopicConnector.java:89]: @mandy-chessell I think I addressed you comment.  I'm not exactly sure which comment you were referring to.

**Comment by jnhagelberg** [comment]
@mandy-chessell Can you please take another look?

**Comment by jnhagelberg** [comment]
@mandy-chessell Can you please take another look?

**Comment by mandy-chessell** [comment]
@jnhagelberg Hello Jeff.  We are still debating what the fix should be.   One of the decisions we made last year was to no longer catch Throwable, and to catch Exception instead.  This would leave the Errors such as the out of memory error to be handled by the JVM and typically cause the process (ie the platform) to exit.  There is concern that if a JVM error occurs, the state of the memory and other aspects of the process is unknown which is why it is not safe to continue.

We are still in the transition and so some of the components are still catching Throwable - unfortunately the Kafka topic connector is one of them.  At one point it even wraps a Throwable with a CheckedException.

This connector clearly needs to be updated so it is no longer catching Throwable.  
Once updated then we need to consider what circumstances are OK to restart the thread and what follow on errors will result.

One scenario is that there is an event that is very big and this causes an error when it is received.  When the thread is restarted, does it:
 * try to receive that event, fail and the thread is restarted again - this could create a loop in one of the servers which is detrimental to the other servers on the platform.
 * the error in the old thread causes the event to be lost.  The recovery of the thread is successful but the event is lost.  In most Egeria situations where federated queries are in operation, the loss of an event is no problem.  But if the repositories are relying on the events then the server needs to be able to log that an event is lost.

Another scenario is that something else in the platform is causing th out of memory exception - at this point the server that the kafka connector belongs to may no longer be viable.

**Comment by planetf1** [comment]
In the out of memory case it may not just be the server that isn't viable but in fact the platform (ie the Java process). In such cases it may be more appropriate for any supervising logic (such as k8s) to restart the platform entirely.

**Comment by jnhagelberg** [comment]
@mandy-chessell @planetf1 I just got back from vacation.  What is the current state of this?  Are there any additional changes I need to make?  It seems like we are in agreement that it can be merged.

**Review by mandy-chessell — APPROVED** [review]

**Comment by jnhagelberg** [comment]
Thanks!!

**Pull request merged**